src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs

```diff
@@ -571,9 +598,33 @@ protected internal void OnAnalyzerConfigDocumentOpened(DocumentId documentId, So
             this.RegisterText(textContainer);
         }
 
+        /// <summary>
+        /// Tries to close the document identified by <paramref name="documentId"/>.  The contents of the closed
+        /// document will be set to whatever the contents of the document are in <see cref="CurrentSolution"/>.
+        /// Subclasses should override this if they want to have different behavior here (for example, reloading the
+        /// contents from disk).
+        /// </summary>
+        /// <param name="documentId"></param>
+        internal virtual void TryOnDocumentClosed(DocumentId documentId)
+        {
+            var doc = this.CurrentSolution.GetDocument(documentId);
```

Comment:
this follows the pattern in most workspaces (adhoc, preview, etc.), which transitino the file to the closed state, but keep the current contents.  The ILspWorkspaces (the only ones that have this be called can override this).

Alternatively, i could just make this throw, and say that you have to override this.  it would only affect the TestWorkspace and any potential mutating lsp server impl in teh future.

--

actually, i like that, i'm going to do that.